### PR TITLE
*: Reorganize Partition DDL states logic | tidb-test=pr/2069

### PR DESCRIPTION
**Pull request by mjonss**
### What problem does this PR solve?

Issue Number: ref #38535

Problem Summary:

### Check List

Tests 

- [ ] Unit test
- [x] Integration test
- [ ] Manual test (add detailed scripts or steps below)
- [ ] No code

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note

Please refer to [Release Notes Language Style Guide](https://pingcap.github.io/tidb-dev-guide/contribute-to-tidb/release-notes-style-guide.html) to write a quality release note.

```release-note
None
```

### CODE REVIEW

File: ddl/partition.go [hunk 1]
```diff
@@ -2164,6 +2188,316 @@ func (w *worker) onExchangeTablePartition(d *ddlCtx, t *meta.Meta, job *model.Jo
...
+			job.State = model.JobStateCancelled
+			return ver, errors.Trace(err)
+		}
+
+		// Re-check that the dropped/added partitions are compatible with current definition
+		firstPartIdx, lastPartIdx, idMap, err := getReplacedPartitionIDs(partNamesCIStr, tblInfo.Partition)
+		if err != nil {
+			return ver, err
+		}
+		var sctx sessionctx.Context
```
Comment (bb7133) [inline ddl/partition.go:2265]: I'm curious if the check `sess == nil` is needed. After all, we used it almost everywhere.

PTAL @wjhuang2016 @tangenta

**Review by bb7133 — APPROVED** [review]
LGTM

File: ddl/partition.go (hunk 1 above)
Comment (wjhuang2016) [reply ddl/partition.go:2265]: No need to check it. It can't be null

File: ddl/ddl_api.go [hunk 2]
```diff
@@ -3897,6 +3897,184 @@ func (d *ddl) AddTablePartitions(ctx sessionctx.Context, ident ast.Ident, spec *
 	return errors.Trace(err)
 }
 
+// getReorganizedDefinitions return the definitions as they would look like after the REORGANIZE PARTITION is done.
+func getReorganizedDefinitions(pi *model.PartitionInfo, firstPartIdx, lastPartIdx int, idMap map[int]struct{}) []model.PartitionDefinition {
```
Comment (wjhuang2016) [inline ddl/ddl_api.go:3901]: The partition can't be the hash partition here?

File: ddl/ddl_api.go [hunk 3]
```diff
@@ -3897,6 +3897,184 @@ func (d *ddl) AddTablePartitions(ctx sessionctx.Context, ident ast.Ident, spec *
...
+	if err = checkReorgPartitionDefs(ctx, meta, partInfo, firstPartIdx, lastPartIdx, idMap); err != nil {
+		return errors.Trace(err)
+	}
+	if err = handlePartitionPlacement(ctx, partInfo); err != nil {
+		return errors.Trace(err)
+	}
+
+	// Do the preSplitAndScatter before the actual DDL work, since it will copy data
+	// and we want to have the regions prepared before that.
+	d.preSplitAndScatter(ctx, meta, partInfo)
```
Comment (wjhuang2016) [inline ddl/ddl_api.go:4000]: It may be strange if the DDL is canceled but the table is splitted and scattered.

File: ddl/partition.go [hunk 4]
```diff
@@ -2164,6 +2188,305 @@ func (w *worker) onExchangeTablePartition(d *ddlCtx, t *meta.Meta, job *model.Jo
...
+
+		err = checkReorgPartitionNames(tblInfo.Partition, partNamesCIStr, partInfo)
+		if err != nil {
+			job.State = model.JobStateCancelled
+			return ver, errors.Trace(err)
+		}
+
+		// Re-check that the dropped/added partitions are compatible with current definition
+		firstPartIdx, lastPartIdx, idMap, err := getReplacedPartitionIDs(partNamesCIStr, tblInfo.Partition)
+		if err != nil {
```
Comment (wjhuang2016) [inline ddl/partition.go:2262]: job.State = model.JobStateCancelled?

File: ddl/partition.go [hunk 5]
```diff
@@ -2164,6 +2188,305 @@ func (w *worker) onExchangeTablePartition(d *ddlCtx, t *meta.Meta, job *model.Jo
...
+			return ver, err
+		}
+		sctx := w.sess.Context
+		if err = checkReorgPartitionDefs(sctx, tblInfo, partInfo, firstPartIdx, lastPartIdx, idMap); err != nil {
+			return ver, err
```
Comment (wjhuang2016) [inline ddl/partition.go:2267]: job.State = model.JobStateCancelled?

File: ddl/ddl_api.go (hunk 2 above)
Comment (tangenta) [reply ddl/ddl_api.go:3901]:
```go
switch pi.Type {
	case model.PartitionTypeRange, model.PartitionTypeList:
	default:
		return errors.Trace(dbterror.ErrUnsupportedReorganizePartition)
	}
```

Comment (mjonss) [reply ddl/ddl_api.go:3901]: HASH based partitioning changes by ADD PARTITION and COALESCE PARTITION, which is not yet supported. But after this feature (REORGANIZE PARTITION) the underlying core is there and we can add the support.

Comment (tiancaiamao) [reply ddl/ddl_api.go:3901]: https://github.com/pingcap/tidb/pull/40473/files#diff-b7c54f7dad7257cbe4fa04b8d76c9bed694e2e2e7c5599f45eb2172d8d4ff8a7R3975

For range and list only

File: ddl/ddl_api.go (hunk 3 above)
Comment (tiancaiamao) [reply ddl/ddl_api.go:4000]: We can count on pd to merge the empty regions.
Pre-scatter bring more goods in general IMO

File: ddl/partition.go (hunk 4 above)
Comment (mjonss) [reply ddl/partition.go:2262]: Great catch, thanks!

File: ddl/partition.go (hunk 5 above)
Comment (mjonss) [reply ddl/partition.go:2267]: Great catch, thanks!

**Review by tiancaiamao — APPROVED** [review]
LGTM, please address comment from the others @mjonss

File: ddl/ddl_api.go (hunk 3 above)
Comment (wjhuang2016) [reply ddl/ddl_api.go:4000]: We can move it before the organization.

Comment (mjonss) [reply ddl/ddl_api.go]: @wjhuang2016 I moved it to the StateNone phase, after the checks.

**Review by wjhuang2016 — APPROVED** [review]

**Pull request merged**